In [ ]:
pip install transformers torch sentencepiece rouge-score sumy nltk


In [ ]:
from transformers import PegasusForConditionalGeneration, PegasusTokenizer

model_name = "google/pegasus-cnn_dailymail"
tokenizer = PegasusTokenizer.from_pretrained(model_name)
model = PegasusForConditionalGeneration.from_pretrained(model_name)

tokenizer_config.json:   0%|          | 0.00/88.0 [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/1.91M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/65.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.12k [00:00<?, ?B/s]

pytorch_model.bin: reconstructing file:   0%|          |  0.00B / 2.28GB            

pytorch_model.bin: downloading bytes:           |  0.00B            

model.safetensors: reconstructing file:   0%|          |  0.00B / 2.28GB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/680 [00:00<?, ?it/s]

[transformers] PegasusForConditionalGeneration LOAD REPORT from: google/pegasus-cnn_dailymail
Key                                  | Status  | 
-------------------------------------+---------+-
model.encoder.embed_positions.weight | MISSING | 
model.decoder.embed_positions.weight | MISSING | 

Notes:
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


generation_config.json:   0%|          | 0.00/280 [00:00<?, ?B/s]

In [ ]:
def summarize(text, num_beams=8, length_penalty=2.0, max_length=100, min_length=30, repetition_penalty=1.2):
    inputs = tokenizer(text, truncation=True, padding="longest", return_tensors="pt")
    summary_ids = model.generate(
        **inputs,
        num_beams=num_beams,
        length_penalty=length_penalty,
        max_length=max_length,
        min_length=min_length,
        no_repeat_ngram_size=3,
        repetition_penalty=repetition_penalty,
        early_stopping=True
    )
    return tokenizer.decode(summary_ids[0], skip_special_tokens=True)

articles = ["""The proposed transaction involves the sale of around a 25% stake in the diagnostics company.

Several global private equity firms, including Warburg Pincus, TA Associates, Bain Capital, and CVC Capital, are reportedly in preliminary discussions to acquire a minority stake in Kochi-based Agappe Diagnostics.

The proposed transaction involves the sale of around a 25% stake in the diagnostics company.

If completed, the deal could value Agappe Diagnostics at approximately INR 3,000-3,200 crore, with the stake sale estimated at INR 700-750 crore. Mauritius-based Sycamore Holdings, through its investment manager Konark Capital, is also reported to be participating in the process.

Founded in 1994, Agappe develops and makes diagnostic reagents and equipment for the global IVD market, with a strong presence in clinical chemistry. The company supplies clinical chemistry, haematology, and immunochemistry products to more than 70,000 laboratories across more than 90 countries.

Currently, Mauritius-based Sycamore Holdings, through its investment manager Konark Capital Partners, holds around 20% in Agappe while the remainder is held by the promoter.

Konark had acquired its stake in 2015 for about INR 50 crore.

In addition, Agappe also operates a wholly owned subsidiary, Agappe Diagnostics Switzerland GmbH, for the international markets. It is also the exclusive Indian distributor for global medical technology companies including Toshiba Medical Systems Corporation (Japan), Mindray (China), and i-SENS (South Korea).

The diagnostic firm reported FY25 revenue of $61 million, EBITDA of $8.7 million, and profit after tax of $3.7 million, according to the latest available data from Tracxn.

Further, the company is expected to clock INR 650 crore in revenue and INR 100-120 crore in EBITDA in FY27.

As per reports, the discussions are at an early stage, and no definitive agreement has been announced. The transaction terms, valuation, and participating investors may change as negotiations progress.

In a separate development, the US-based private equity firm Warburg Pincus is in advanced talks to acquire a key business unit of Maneesh Pharmaceuticals in a transaction valued at INR 1,600 crore to INR 1,800 crore, reflecting the firm's continued interest in India's healthcare sector.
""", """ The greatest footballer of all-time and the heir to his crown on the right wing at Barcelona will face each other for the first time when Argentina play Spain in Sunday's World Cup final.
Given Lionel Messi is now 39 and Lamine Yamal just 19, it is testament to the former's relentless perseverance and the latter's precocious talent that a match featuring both can take place at all, never mind in the sport's greatest fixture.
But this is far from the first time the two have crossed paths.
That happened in 2007, when a 20-year-old Messi had recently become a regular starter for Barcelona, and Lamine Yamal was only five months old.
The meeting is documented in a series of photographs taken by Joan Monfort.
They resurfaced two years ago, when Lamine Yamal helped Spain win Euro 2024, as his father posted one online with the caption "The beginning of two legends".
In the photos, a smiling Messi is cradling and bathing a tiny baby boy who would inexplicably follow so closely in his footballing footsteps.
"It is a true miracle of destiny", Monfort tells BBC Sport.
"It is serendipity - when you find something extra special, so much bigger than you ever thought.
"If you wrote this in a film it would not seem possible."
'Messi had no idea what to do with a baby'
The photoshoot took place in the away dressing room at Barcelona's Camp Nou stadium.
Lamine Yamal's family had won a competition to be there.
His parents, Morocco-born Mounir Nasraoui and Sheila Ebana from Equatorial Guinea, met after moving to Catalonia as children with their own families.
Lamine Yamal is their eldest child, and shortly after he was born, they entered a raffle run by Catalan newspaper Sport in conjunction with Barcelona shirt sponsor and international children's charity Unicef.
Those selected at random would have professional photos taken of their baby with a Barcelona first-team player.
Lamine Yamal was among the winners, and on the day the family arrived, they were paired with Messi by chance.
""", """ OnePlus arrived on the scene in 2014 with brash marketing and a compelling pitch: What if your phone was cheaper and faster? More than a decade later, the market is much different, and so is OnePlus. Confirming months of rumors and speculation, OnePlus has confirmed it's ending phone releases in North America and Europe.

This development has been expected for a while. After a brief period of expansion in the US through deals with T-Mobile and Verizon, OnePlus began visibly shifting its focus to India after the pandemic. OnePlus has also increasingly come under the control of its parent company, Oppo, aligning its device releases and software experience with the larger brand. Still, OnePlus has continued to launch flagship devices in the US and Europe, with the OnePlus 15 going on sale in late 2025. That will be the last OnePlus phone many people see, though.

OnePlus denied rumors of a shutdown earlier this year, but its language was vague and never promised future phone releases. The new statement is also worded gingerly, but the outcome is clear. "As part of the proactive global strategy adjustment, OnePlus has decided to conclude new product rollouts in Europe and North America," the official post reads.

You can still buy a current-gen OnePlus phone in North America and Europe, and the company will offer software support. However, OnePlus devices will drop the OxygenOS build of Android, adopting Oppo's ColorOS with the Android 17 update globally. Only India and China will see new devices—at least for now. A shutdown of the OEM's India operations in 2027 has also been rumored.
"""]  # your text strings

for i, art in enumerate(articles):
    summary = summarize(art)
    print(f"Article {i+1}")
    print(f"Original words: {len(art.split())}")
    print(f"Summary words: {len(summary.split())}")
    print(f"Summary: {summary}\n")

Article 1
Original words: 340
Summary words: 44
Summary: Several global private equity firms, including Warburg Pincus, TA Associates, Bain Capital, and CVC Capital, are reportedly in preliminary discussions to acquire a minority stake in Agappe Diagnostics .<n>The proposed transaction involves the sale of around a 25% stake in the diagnostics company .

Article 2
Original words: 341
Summary words: 47
Summary: Argentina's Lionel Messi and Spain's Lamine Yamal will face each other in the World Cup final on Sunday .<n>The meeting is documented in a series of photographs taken by Joan Monfort.<n>It happened in 2007, when a 20-year-old Messi had recently become a regular starter for Barcelona .

Article 3
Original words: 266
Summary words: 37
Summary: OnePlus has confirmed it's ending phone releases in North America and Europe .<n>The company has shifted its focus to India after a brief period of expansion in the U.S.<n>Only India and China will see new devices .



In [ ]:
# num_beams comparison
for beams in [1, 4, 8]:
    print(f"num_beams={beams}:", summarize(articles[2], num_beams=beams))

# length_penalty comparison
for lp in [0.8, 1.0, 2.0]:
    print(f"length_penalty={lp}:", summarize(articles[2], length_penalty=lp))

[transformers] The following generation flags are not valid and may be ignored: ['early_stopping', 'length_penalty']. Set `TRANSFORMERS_VERBOSITY=info` for more details.


num_beams=1: OnePlus is ending phone releases in North America and Europe .<n>The move was expected, but the company has been shifting its focus to India .
num_beams=4: OnePlus has confirmed it's ending phone releases in North America and Europe .<n>The company has increasingly come under the control of its parent company, Oppo.<n>Only India and China will see new devices .
num_beams=8: OnePlus has confirmed it's ending phone releases in North America and Europe .<n>The company has shifted its focus to India after a brief period of expansion in the U.S.<n>Only India and China will see new devices .
length_penalty=0.8: OnePlus has confirmed it's ending phone releases in North America and Europe .<n>The company has increasingly come under the control of its parent company, Oppo .
length_penalty=1.0: OnePlus has confirmed it's ending phone releases in North America and Europe .<n>The company has shifted its focus to India after a brief period of expansion in the U.S.<n>Only India and Chin

In [ ]:
!pip install rouge-score -q

In [ ]:

from rouge_score import rouge_scorer
import pandas as pd

# ---- Manual reference (human) summaries for the 3 articles ----
reference_summaries = [
    "Global PE firms are in early talks to buy a 25% stake in Agappe Diagnostics, "
    "valuing the Kochi-based company at around INR 3,000-3,200 crore.",

    "Messi and Lamine Yamal, who will face off in the World Cup final, first met "
    "as adults and baby in 2007 when Messi held a newborn Yamal at Camp Nou.",

    "OnePlus is ending new phone launches in North America and Europe, shifting "
    "its focus mainly to India and China."
]

# ---- Generate Pegasus summaries ----
generated_summaries = [summarize(art) for art in articles]

# ---- Compute ROUGE scores ----
scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)

results = []
for i, (ref, gen) in enumerate(zip(reference_summaries, generated_summaries)):
    scores = scorer.score(ref, gen)
    results.append({
        "Article": f"Article {i+1}",
        "ROUGE-1": round(scores['rouge1'].fmeasure, 3),
        "ROUGE-2": round(scores['rouge2'].fmeasure, 3),
        "ROUGE-L": round(scores['rougeL'].fmeasure, 3),
    })

df_rouge = pd.DataFrame(results)
print(df_rouge)

     Article  ROUGE-1  ROUGE-2  ROUGE-L
0  Article 1    0.389    0.143    0.333
1  Article 2    0.386    0.222    0.386
2  Article 3    0.533    0.310    0.500


In [ ]:
# ---- Present ROUGE scores in a table with averages ----
df_display = df_rouge.copy()
df_display.loc['Average'] = [
    'Average',
    round(df_rouge['ROUGE-1'].mean(), 3),
    round(df_rouge['ROUGE-2'].mean(), 3),
    round(df_rouge['ROUGE-L'].mean(), 3)
]
df_display

     Article  ROUGE-1  ROUGE-2  ROUGE-L
0  Article 1    0.389    0.143    0.333
1  Article 2    0.386    0.222    0.386
2  Article 3    0.533    0.310    0.500


### Interpretation of ROUGE Scores

| Article | ROUGE-1 | ROUGE-2 | ROUGE-L |
|---|---|---|---|
| Article 1 (Agappe Diagnostics) | 0.389 | 0.143 | 0.333 |
| Article 2 (Messi & Lamine Yamal) | 0.386 | 0.222 | 0.386 |
| Article 3 (OnePlus) | 0.533 | 0.310 | 0.500 |
| **Average** | **0.436** | **0.225** | **0.406** |

**Observations:**

- **ROUGE-1 > ROUGE-L > ROUGE-2 for every article.** This pattern is typical: matching single words (ROUGE-1) is much easier than matching sequences of two consecutive words (ROUGE-2), so ROUGE-2 is consistently the lowest score. ROUGE-L (longest common subsequence) sits in between since it rewards in-order overlap without requiring strict adjacency.

- **Article 3 (OnePlus) scored highest across all three metrics** (ROUGE-1 = 0.533, ROUGE-2 = 0.310, ROUGE-L = 0.500). The Pegasus summary for this article stuck closely to the same key facts and wording used in the reference summary (ending phone launches, shift to India/China), producing strong word- and phrase-level overlap.

- **Article 1 (Agappe Diagnostics) scored lowest on ROUGE-2 and ROUGE-L** (0.143 and 0.333). Even though the summary covers the same general topic (PE stake sale), Pegasus likely phrased the numbers/entities differently or emphasized different details than the human reference, reducing the amount of matching multi-word sequences.

- **Overall**, the moderate-to-good ROUGE scores (averaging ~0.44 for ROUGE-1) suggest Pegasus is capturing the core topic and several key terms from each article, but — as an *abstractive* summarizer — it doesn't copy phrasing verbatim from a human-style reference, so perfect overlap is neither expected nor achieved. Higher ROUGE-2/ROUGE-L scores (like Article 3) indicate summaries that are more extractive in style, while lower scores (like Article 1) indicate more paraphrasing/re-wording by the model.

In [ ]:
!pip install sumy nltk -q
import nltk
nltk.download('punkt')
nltk.download('punkt_tab')

In [ ]:
# ---- Q9: Extractive summarization with sumy's TextRank, on the same 3 articles ----
from sumy.parsers.plaintext import PlaintextParser
from sumy.nlp.tokenizers import Tokenizer
from sumy.summarizers.text_rank import TextRankSummarizer

def textrank_summarize(text, sentence_count=3):
    parser = PlaintextParser.from_string(text, Tokenizer("english"))
    summarizer = TextRankSummarizer()
    summary = summarizer(parser.document, sentence_count)
    return " ".join(str(s) for s in summary)

textrank_summaries = [textrank_summarize(art) for art in articles]

for i, ts in enumerate(textrank_summaries):
    print(f"Article {i+1}")
    print(f"TextRank Summary: {ts}\n")

Article 1
TextRank Summary: The proposed transaction involves the sale of around a 25% stake in the diagnostics company. Further, the company is expected to clock INR 650 crore in revenue and INR 100-120 crore in EBITDA in FY27. In a separate development, the US-based private equity firm Warburg Pincus is in advanced talks to acquire a key business unit of Maneesh Pharmaceuticals in a transaction valued at INR 1,600 crore to INR 1,800 crore, reflecting the firm's continued interest in India's healthcare sector.

Article 2
TextRank Summary: The greatest footballer of all-time and the heir to his crown on the right wing at Barcelona will face each other for the first time when Argentina play Spain in Sunday's World Cup final. Given Lionel Messi is now 39 and Lamine Yamal just 19, it is testament to the former's relentless perseverance and the latter's precocious talent that a match featuring both can take place at all, never mind in the sport's greatest fixture. In the photos, a smiling 

In [ ]:
# ---- Compute ROUGE for TextRank summaries against the same reference summaries, and compare to Pegasus ----
textrank_results = []
for i, (ref, gen) in enumerate(zip(reference_summaries, textrank_summaries)):
    scores = scorer.score(ref, gen)
    textrank_results.append({
        "Article": f"Article {i+1}",
        "TextRank_ROUGE-1": round(scores['rouge1'].fmeasure, 3),
        "TextRank_ROUGE-2": round(scores['rouge2'].fmeasure, 3),
        "TextRank_ROUGE-L": round(scores['rougeL'].fmeasure, 3),
    })

df_textrank = pd.DataFrame(textrank_results)

# Merge Pegasus and TextRank scores side by side
df_compare = df_rouge.rename(columns={
    "ROUGE-1": "Pegasus_ROUGE-1", "ROUGE-2": "Pegasus_ROUGE-2", "ROUGE-L": "Pegasus_ROUGE-L"
}).merge(df_textrank, on="Article")

print(df_compare)

     Article  Pegasus_ROUGE-1  Pegasus_ROUGE-2  Pegasus_ROUGE-L  TextRank_ROUGE-1  TextRank_ROUGE-2  TextRank_ROUGE-L
0  Article 1            0.389            0.143            0.333             0.298             0.071             0.175
1  Article 2            0.386            0.222            0.386             0.299             0.091             0.164
2  Article 3            0.533            0.310            0.500             0.298             0.109             0.170


### Q10: Reflection — Pegasus (Abstractive) vs TextRank (Extractive)

**Differences observed between abstractive and extractive summaries**

- Pegasus (abstractive) generated **new sentences** that paraphrase and compress the article — e.g., for Article 3 it wrote *"OnePlus has confirmed it's ending phone releases in North America and Europe"*, condensing several paragraphs into one clean sentence not found verbatim in the source.
- TextRank (extractive) simply **selected 3 existing sentences** from the article ranked as most "central" by sentence-similarity graph scoring, and stitched them together unchanged. As a result, TextRank summaries read less fluently as a standalone summary — they can feel like disconnected excerpts rather than a coherent narrative (most noticeable in Article 2, where TextRank pulled the opening context sentence and a mid-article visual description, skipping the connecting facts about *when* and *where* Messi and Yamal met).
- Pegasus summaries were noticeably **shorter and denser** (37–47 words) than the TextRank summaries (which, being made of full original sentences, ran much longer per summary), showing Pegasus's stronger compression ability.
- Interestingly, **Pegasus scored higher ROUGE-1/2/L on all three articles** than TextRank against our human references. This is somewhat counter-intuitive since ROUGE rewards *n*-gram overlap and extractive methods copy exact wording — but here the human reference summaries were themselves written abstractively (paraphrased, short), so they matched Pegasus's compressed phrasing better than TextRank's long, verbatim sentence extracts.

**Factual inconsistencies / hallucinations noticed in Pegasus output**

- In Article 1, Pegasus's summary drops the word "Kochi-based" and the specific valuation figures (INR 3,000–3,200 crore) that were central to the article — not a hallucination, but an **omission of key facts**, which can be misleading if a reader only sees the summary.
- In Article 2, Pegasus states the meeting *"happened in 2007, when a 20-year-old Messi had recently become a regular starter for Barcelona"* — this is factually consistent with the source, but the summary **loses the important detail that Yamal was a 5-month-old baby**, which was the whole point of the story; without that context the summary reads oddly (an adult meeting is implied, when it was a baby photo shoot).
- No outright fabricated facts (invented names, numbers, or events) were observed in these three examples, but Pegasus's tendency to **drop qualifying details** (dates, ages, specific entities) is a mild form of unfaithfulness — the summary can be technically accurate yet misleading by omission.

**When to prefer extractive vs abstractive summarization**

- **Prefer extractive (TextRank)** when factual accuracy and traceability are critical — e.g., legal documents, medical records, financial disclosures — since every sentence in the summary can be traced back verbatim to the source, eliminating hallucination risk.
- **Prefer extractive** also when compute budget is limited or a lightweight, fast, non-neural solution is needed (TextRank requires no GPU/large model download).
- **Prefer abstractive (Pegasus)** when a short, fluent, human-readable summary is the priority — e.g., news digest apps, headline generation, chatbot summarization — where a natural sentence is more useful than a stitched-together excerpt.
- **Prefer abstractive** when the source text is long and repetitive, since it can compress much more aggressively than extractive methods (which are limited to whole original sentences).
- A hybrid approach (extractive pass to identify key sentences, then abstractive rewriting) is often used in production systems to combine faithfulness with fluency.